aten/src/TH/generic/THTensorRandom.cpp

```diff
@@ -247,9 +247,9 @@ void THTensor_(multinomialAliasDraw)(THLongTensor *self, THGenerator *_generator
 
       J_sample = THLongTensor_fastGet1d(J, rand_ind);
 
-      sample_idx = J_sample*(1 -_mask) + (rand_ind+1L) * _mask;
+      sample_idx = J_sample*(1 -_mask) + rand_ind * _mask;
 
-      THLongTensor_fastSet1d(self, i, sample_idx-1L);
+      THLongTensor_fastSet1d(self, i, sample_idx);
```

Comment:
Good catch. I wonder if this was put in place because `J` was mistakenly thought to be 1 indexed instead of 0 indexed?

Just one comment based on this. In `multinomial_alias_setup`, towards the end, we do the following. I do not fully understand when this condition gets triggered, but if J can have a 0 index, this looks wrong and we probably need to change the if block to `J_data[i] < 0`. An easy way to see this would be to use a test case where the bulk of the probability mass is on the first index, e.g. probs = `torch.tensor([0.8, 0.2])` and see if your test passes.

```
    {
      // sometimes an large index isn't added to J.	      // sometimes an large index isn't added to J.
      // fix it by making the probability 1 so that J isn't indexed.	      // fix it by making the probability 1 so that J isn't indexed.
      if(J_data[i] <= 0)	      if(J_data[i] <= 0)
        q_data[i] = 1.0;	        q_data[i] = 1.0;
    }	    }
```

Reply:
Oh..I think the reason why J is 1-indexed is because the 0 index is used (or at least seems to have been meant) to designate the null case, i.e. the case where there is no alias. e.g. for the case above where `probs = [0.8, 0.2]`, `J = [0, 1]` and `q = [1., 0.4]`. That is to say that if `rand_ind` is 0, we always return `0` and if `rand_ind` is 1, we return `0` with probability 0.6 and `1` with probability 0.4. 

In any case, you'll need the check above to set q correctly and can probably decrement J if needed, but it might just be easier to follow the earlier convention of 1-indexing. The bug is probably at some other place in that case.

Reply:
Well, it seems like for `probs = [0.8, 0.2]`, `J` is `[0, 0]` (not `[0, 1]`), and `q` is `[1., 1.]` (not `[1., 0.4]`). 

I report back with more observations (my local build got botched due to a rebase).

Reply:
I replaced `<=` with `<`, and the test case passes. However, `J` is still `[0, 0]`.

Reply:
I think J = [0, 1] with the previous 1-indexed logic. It doesn't really matter what J[0] is really, because it will never be selected if q[0] = 1. Let me take another look at your changes.